src/torchcodec/_frame.py

```diff
@@ -57,9 +63,52 @@ class FrameBatch(Iterable):
     duration_seconds: Tensor
     """The duration of the frame, in seconds (1-D ``torch.Tensor`` of floats)."""
 
-    def __iter__(self) -> Iterator[Union[Tensor, float]]:
-        for field in dataclasses.fields(self):
-            yield getattr(self, field.name)
+    def __post_init__(self):
+        if self.data.ndim < 4:
+            raise ValueError(
+                f"data must be at least 4-dimensional. Got {self.data.shape = } "
+                "For 3-dimensional data, create a Frame object instead."
+            )
+
+        leading_dims = self.data.shape[:-3]
+        if not (leading_dims == self.pts_seconds.shape == self.duration_seconds.shape):
+            raise ValueError(
+                "Tried to create a FrameBatch but the leading dimensions of the inputs do not match. "
+                f"Got {self.data.shape = } so we expected the shape of pts_seconds and "
+                f"duration_seconds to be {leading_dims = }, but got "
+                f"{self.pts_seconds.shape = } and {self.duration_seconds.shape = }."
+            )
+
+    def __iter__(self) -> Union[Iterator["FrameBatch"], Iterator[Frame]]:
+        cls = Frame if self.data.ndim == 4 else FrameBatch
+        for data, pts_seconds, duration_seconds in zip(
+            self.data, self.pts_seconds, self.duration_seconds
+        ):
+            yield cls(
+                data=data,
+                pts_seconds=pts_seconds,
+                duration_seconds=duration_seconds,
+            )
+
+    def __getitem__(self, key) -> Union["FrameBatch", Frame]:
+        data = self.data[key]
+        pts_seconds = self.pts_seconds[key]
+        duration_seconds = self.duration_seconds[key]
+        if self.data.ndim == 4:
+            return Frame(
+                data=data,
+                pts_seconds=float(pts_seconds.item()),
+                duration_seconds=float(duration_seconds.item()),
+            )
+        else:
+            return FrameBatch(
+                data=data,
+                pts_seconds=pts_seconds,
+                duration_seconds=duration_seconds,
+            )
```

Comment:
Tensor has a .item() method for returning the underlying dtype.

Should we have something like that here? i.e. always return a FrameBatch but return a Frame if .item() is called?

Reply:
> always return a FrameBatch but return a Frame if .item() is called?

I feel like this is what we're already doing, but perhaps I'm misunderstanding?

BTW, this quirk is only needed for mypy (sigh). Originally the code was just:

```py
        cls = Frame if self.data.ndim == 4 else FrameBatch
        return cls(
            self.data[key],
            self.pts_seconds[key],
            self.duration_seconds[key],
        )
```

and everything was fine, and the Frame would get proper float value because of what we do in its `post_init`. But mypy was complaining so I had to go for this in [`4661237` (#283)](https://github.com/pytorch/torchcodec/pull/283/commits/46612374677ff4dbbea5fdf2b175bc1090c4d1fa)

Reply:
> I feel like this is what we're already doing, but perhaps I'm misunderstanding?

Don't we return a Frame for the special case of dimensions=4?

What I am saying is we should return a FrameBatch even in that case (of size 1). So we are consistent with Tensor

I'll leave it to you

Reply:
I don't have a super strong preference on this - let me open an issue so we can discuss during one of the meetings